crt_portal/cts_forms/mail.py

```diff
@@ -1,17 +1,52 @@
-from django.core.mail import send_mail
-from django.conf import settings
 import logging
+from datetime import datetime
+
+from django.conf import settings
+from django.core.mail import send_mail
+from tms.models import TMSEmail
 
 logger = logging.getLogger(__name__)
 
 
+def remove_disallowed_recipients(recipient_list):
+    """
+    If we've limited the recipients, modify the `to` attribute of our email_message
+    """
+    # If the list of restrictions is empty, allow all
+    restricted_to = settings.RESTRICT_EMAIL_RECIPIENTS_TO
+    if restricted_to:
+        recipient_list = [
+            to_address
+            for to_address in recipient_list
+            if to_address in restricted_to
+        ]
+    return recipient_list
+
+
 def crt_send_mail(report, template):
     """
     Given a report and a template, use django's builtin `send_mail` to generate and send
     an outbound email
+
+    Returns a reference to the delivered TMSAPI request
     """
     subject = template.render_subject(report)
     message = template.render_body(report)
-    recipient_list = [report.contact_email]
-    send_mail(subject, message, settings.DEFAULT_FROM_EMAIL, recipient_list, fail_silently=False)
+
+    recipient_list = remove_disallowed_recipients([report.contact_email])
+    if not recipient_list:
+        logger.info(f'{report.contact_email} not in allowed domains, not attempting to deliver email response template #{template.id} to report: {report.id}')
+        return None
+    send_results = send_mail(subject, message, settings.DEFAULT_FROM_EMAIL, recipient_list, fail_silently=False)
     logger.info(f'Sent email response template #{template.id} to report: {report.id}')
+    if settings.EMAIL_BACKEND == 'tms.backend.TMSEmailBackend':
+        response = send_results[0]
+        TMSEmail(tms_id=response['id'],
+                 recipient=report.contact_email,
+                 subject=subject,
+                 body=message,
+                 report=report,
+                 created_at=datetime.strptime(response['created_at'], '%Y-%m-%dT%H:%M:%S%z'),
+                 status=response['status']
+                 ).save()
```

Comment:
My gut reaction was that this bit feels like it should live in the `tms` app - thoughts?

I'm guessing at least part of the reason it's here is so that we can reference the `report`

Reply:
I struggled with this part a bit and couldn't settle on a cleaner approach. I think I got stuck because `TMSEmail` is too coupled to the TMS backend but in a perfect world we want to track/record outbound emails from any backend. Like in the future if we migrate away from TMS. This combined with `send_mail` returning different things based on the backend, makes this less than ideal and a bit confusing for sure. 

I'd love to pair and look at it together if possible!

Reply:
Right. I think most/all of those fields would map pretty cleanly onto what we'd track for other email backends (`tms_id` could become `external_id` or something to be more generic). I'm wondering if the backend should do the work of normalizing this data into the shape of some generic `OutboundEmail` model  that could feasibly be reused if we switched providers in the future